libs/vm/src/module.cpp

```diff
@@ -143,72 +149,122 @@ double toFloat64(VM * /* vm */, AnyPrimitive const &from)
   return Cast<double>(from);
 }
 
+class AllClassInterfaces
```

Comment:
It feels like this is unnecessarily complex.
The `ClassInterface<T>` design supports builder design pattern, which allows to fully setup the interface exclusivell using r-value (= no necessity to create l-value variable).

Reason being, that instead of unnecessarily creating l-value (represented as class data member) and then passing it around, the following builder patter nicely simplifies the thing:

```
  GetClassInterface<IArray>()
    .CreateConstuctor<int32_t>()
    .CreateMemberFunction("count", &IArray::Count)
    .CreateMemberFunction("append", &IArray::Append)
    .CreateMemberFunction("popBack", &IArray::PopBackOne)
    .CreateMemberFunction("popBack", &IArray::PopBackMany)
    .CreateMemberFunction("popFront", &IArray::PopFrontOne)
    .CreateMemberFunction("popFront", &IArray::PopFrontMany)
    .CreateMemberFunction("reverse", &IArray::Reverse)
    .CreateMemberFunction("extend", &IArray::Extend)
    .EnableIndexOperator<AnyInteger, TemplateParameter>()
    .CreateInstantiationType<Array<bool>>()
    .CreateInstantiationType<Array<int8_t>>()
    .CreateInstantiationType<Array<uint8_t>>()
    .CreateInstantiationType<Array<int16_t>>()
    .CreateInstantiationType<Array<uint16_t>>()
    .CreateInstantiationType<Array<int32_t>>()
    .CreateInstantiationType<Array<uint32_t>>()
    .CreateInstantiationType<Array<int64_t>>()
    .CreateInstantiationType<Array<uint64_t>>()
    .CreateInstantiationType<Array<float>>()
    .CreateInstantiationType<Array<double>>()
    .CreateInstantiationType<Array<Ptr<String>>>();
```

Reply:
Context for this change: this came out of my conversation with @robertdickson yesterday. I think the current arrangement is not extensible, as it interleaves registering new type instantiations and member functions that may need certain types to have been registered with the VM. For example, with the current shape of the code on `master`, it's impossible to add a `String.split(String) : Array<String>` member function without moving the whole String block below the Array block, where `Array<String>` is instantiated. This kind of connascence is unnecessary, and can be avoided by registering types up front.

The proposed Builder pattern, while pretty, does not solve the problem relevant to this PR.

Reply:
You can call `GetClassInterface<Array>()` at any point, disregard whether the `Array` type has been previously setup or not, for example, the following should work just fine:

```cpp
//This can be called anywhere (even before preceding `Array` full/partial setup), and is sort-of equivalent of C++ forward declaration of the class type.
GetClassInterface<IArray>();

//Then you can call do of the `String` type:
GetClassInterface<String>().CreateMemberFunction("split", &String::Split);

...

//Then sometime later you can finish/extend setup of `Array` type:
GetClassInterface<IArray>()
    .CreateConstuctor<int32_t>()
    .CreateMemberFunction("count", &IArray::Count)
    .CreateMemberFunction("append", &IArray::Append)
    .CreateMemberFunction("popBack", &IArray::PopBackOne)
    .CreateMemberFunction("popBack", &IArray::PopBackMany)
    . ...
```

Reply:
That's not true I'm afraid. The IArray type has been created before, inside the Analyser.

Reply:
You can call `GetClassInterface<Array>().CreateConstuctor()` at any point, disregard of whether the `Array` type has been previously setup or not, what makes it behave as sort-of equivalent of C++ *forward declaration* of the class type.

For example, the following should work just fine:

```cpp
//This can be called anywhere (even before preceding `Array` full/partial setup),
// and is sort-of equivalent of C++ forward declaration of the class type.
GetClassInterface<Array>().CreateConstuctor();

//Then you can call do of the `String` type:
GetClassInterface<String>().CreateMemberFunction("split", &String::Split);

...

//Then sometime later you can finish/extend setup of `Array` type:
GetClassInterface<IArray>()
    .CreateConstuctor<int32_t>()
    .CreateMemberFunction("count", &IArray::Count)
    .CreateMemberFunction("append", &IArray::Append)
    .CreateMemberFunction("popBack", &IArray::PopBackOne)
    .CreateMemberFunction("popBack", &IArray::PopBackMany)
    . ...
```

Reply:
That's not true I'm afraid. The IArray type has been created before, inside the Analyser.

Reply:
I would rather introduce Small change in to the `Module::ClassInterface<T>` class, which would enable registration of the type without necessity/forcing to add methods to that type. This would be sort-of equivalent to the C++ forward declaration.
This can be implemented directly in the constructor of the `Module::ClassInterface<T>` class (what would make it nice to declare the type by simply calling `GetClassInterface<T>()` function, or we can put this in to separate method of the class to make it explicit, for instance `Module::ClassInterface<T>::ForwardDeclare()` method (so one would need to call the `GetClassInterface<T>().ForwardDeclare()`).

Once the above would be done then you can simply `GetClassInterface<Array>()` at any point, disregard of whether the `Array` type has been previously setup or not, for example, the following should work just fine:

```cpp
//This can be called anywhere (even before preceding `Array` full/partial setup),
// and is sort-of equivalent of C++ forward declaration of the class type.
GetClassInterface<Array>();

//Then you can call do of the `String` type:
GetClassInterface<String>().CreateMemberFunction("split", &String::Split);

...

//Then sometime later you can finish/extend setup of `Array` type:
GetClassInterface<IArray>()
    .CreateConstuctor<int32_t>()
    .CreateMemberFunction("count", &IArray::Count)
    .CreateMemberFunction("append", &IArray::Append)
    .CreateMemberFunction("popBack", &IArray::PopBackOne)
    .CreateMemberFunction("popBack", &IArray::PopBackMany)
    . ...
```

Reply:
You can register a type before defining its methods -- that's creating the class! Module has CreateClassType(). Template creation is not, by design, available to the Module system as it is private to the compiler.

Reply:
Core primitives, String and the templates are created inside the compiler and are thus already registered.

Reply:
The problem with `Module::CreateClassType(std::string const &name)` method is that it requires the `name` string parameter.
I was looking in to this method originally, and that method felt like good candidate to start off with to implement that _forward declaration_ concept I mentioned above. In this sense I keep advocating to create separate dedicated method called `ForwardDeclare()` (or something like that) in orde rto signify, that programmer can call that method without fear that it will do something unexpected behind the scenes (e.g. registering the same type multiple times, etc ...)